File: algorithms/easy/count_days_spent_together.cpp
class Solution {
public:
  int dateToDays(string s) {
    vector<int> days{31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
    int m = 0;
    int d = 0;
    sscanf(s.c_str(), "%d-%d", &m, &d);
    int ds = accumulate(days.begin(), days.begin() + m - 1, 0);
    return ds + d;
  }
  int countDaysTogether(string arriveAlice, string leaveAlice, string arriveBob, string leaveBob) {
    int alice_start = dateToDays(arriveAlice);
    int alice_end = dateToDays(leaveAlice);
    int bob_start = dateToDays(arriveBob);
    int bob_end = dateToDays(leaveBob);
    return max(0, min(alice_end, bob_end) - max(alice_start, bob_start) + 1);
  }
};

File: algorithms/easy/count_days_spent_together.ex
defmodule Solution do
  @spec count_days_together(
          arrive_alice :: String.t(),
          leave_alice :: String.t(),
          arrive_bob :: String.t(),
          leave_bob :: String.t()
        ) :: integer
  def count_days_together(arrive_alice, leave_alice, arrive_bob, leave_bob) do
    alice_start = date_to_days(arrive_alice)
    alice_end = date_to_days(leave_alice)
    bob_start = date_to_days(arrive_bob)
    bob_end = date_to_days(leave_bob)
    max(0, min(alice_end, bob_end) - max(alice_start, bob_start) + 1)
  end

  def date_to_days(s) do
    days = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]

    s
    |> String.split("-")
    |> Enum.map(&String.to_integer/1)
    |> then(fn [m, d] -> d + (Enum.take(days, m - 1) |> Enum.sum()) end)
  end
end
